fractal_tasks_core/lib_masked_loading.py

```diff
@@ -0,0 +1,248 @@
+            highres_region=region,
+            highres_shape=current_label_array.shape,
+            lowres_shape=masking_label_array.shape,
+        )
+        masking_label_region = masking_label_array[lowres_region].compute()
+        masking_label_region = upscale_array(
+            array=masking_label_region,
+            target_shape=current_label_region.shape,
+        )
+    else:
+        masking_label_region = masking_label_array[region].compute()
+
+    # Check that all shapes match
+    shapes = (
+        masking_label_region.shape,
+        current_label_region.shape,
+        image_array.shape[1:],
+    )
+    if len(set(shapes)) > 1:
+        raise ValueError(
+            "Shape mismatch:\n"
+            f"{current_label_region.shape=}\n"
+            f"{masking_label_region.shape=}\n"
+            f"{image_array.shape=}"
+        )
+
+    # Compute background mask
+    background_3D = masking_label_region != label_value
+    if (masking_label_region == label_value).sum() == 0:
+        raise ValueError(
+            f"Label {label_value} is not present in the extracted ROI"
+        )
+
+    # Set image background to zero
+    background_4D = np.expand_dims(background_3D, axis=0)
+    image_array[background_4D] = 0
+
+    return (image_array, background_3D, current_label_region)
+
+
+def postprocess_cellpose_output(
+    *,
+    modified_array: np.ndarray,
+    original_array: np.ndarray,
+    background: np.ndarray,
+) -> np.ndarray:
+    """
+    Postprocess a cellpose input, mainly to restore its original background
+
+    **FIXME 1**: review/improve variable names
+
+    **FIXME 2**: make this function more flexible, and move it to a separate
+    module (see `issue 340
+    <https://github.com/fractal-analytics-platform/fractal-tasks-core/issues/340>`_).
+
+    :param modified_array: The 3D (ZYX) array with the correct object data and
+                           wrong background data.
+    :param original_array: The 3D (ZYX) array with the wrong object data and
+                           correct background data.
+    :param background: The 3D (ZYX) boolean array that defines the
+                       background.
+    """
+
+    # Restore background
+    modified_array[background] = original_array[background]
+    return modified_array
+
+
+def wrap_function_with_masking(
+    *,
+    function: Callable,
+    image_array: np.ndarray,
+    kwargs: Optional[dict] = None,
+    use_masks: bool,
+    preprocessing_kwargs: Optional[dict] = None,
+    postprocessing_kwargs: Optional[dict] = None,
+):
+    """
+    Wrap a function with some pre/post-processing functions
+
+    :param function: The callable function to be wrapped.
+    :param args: Positional arguments for ``function``.
+    :param kwargs: Keyword arguments for ``function``.
+    :param use_masks: If ``False``, the wrapper only calls ``function(*args,
+                      **kwargs)``.
+    :param preprocessing_kwargs: Keyword arguments for the preprocessing
```

Comment:
Would be good to either list the required arguments (region, current_label_path, ROI_table_path, ROI_index are all required, right?) or point to the actual function above (`preprocess_cellpose_input`) and check that they are actually provided (otherwise, the preprocessing function will fail with confusing error messages.

Reply:
I'll clarify that, but there is a more general discussion about this to be had in #340.
Introducing a wrapper is not for free, since we suddenly loose the clear view of "this is the preprocessing function, these are its parameters", and everything is mediated via the wrapper.

The current masked-loading code is a complicated way of doing what was already done a few commits ago (i.e. without wrapper), but we're not yet at the point where this wrapper would be generally useful (e.g. for other tasks). That is part of #340.

Reply:
Ok, sounds good, let's continue any discussions there :)

Reply:
Ok, sounds good, let's continue any discussions there :)

Reply:
* I added a reference to the preprocessing function in the docstring of the wrapper.
* I fixed an error in the preprocessing-function call signature, which still had `None` defaults for its arguments (since in previous versions it also included the `use_masks=False` branch, which is now handled in the wrapper). This change makes all kwargs required, so that the error messages if the user forgets a kwargs when calling the wrapper now looks reasonable:
```python traceback
Traceback (most recent call last):
  File "/home/tommaso/Fractal/fractal-tasks-core/examples/09_secondary_labeling/02_cellpose_two_steps_segmentation.py", line 48, in <module>
    cellpose_segmentation(
  File "/home/tommaso/Fractal/fractal-tasks-core/fractal_tasks_core/cellpose_segmentation.py", line 542, in cellpose_segmentation
    new_label_img = masked_loading_wrapper(
  File "/home/tommaso/Fractal/fractal-tasks-core/fractal_tasks_core/lib_masked_loading.py", line 237, in masked_loading_wrapper
    ) = _preprocess_input(image_array, **preprocessing_kwargs)
TypeError: _preprocess_input() missing 1 required keyword-only argument: 'ROI_positional_index'
```